src/server/pkg/storage/chunk/util.go

```diff
@@ -36,3 +39,40 @@ func RandSeq(n int) []byte {
 	}
 	return []byte(string(b))
 }
+
+// CopyN is an efficient copy function that turns full chunk copies into data reference writes.
+func CopyN(w *Writer, r *Reader, n int64) error {
+	if r.Len() < n {
+		// (bryce) this should be an internal error.
+		return fmt.Errorf("number of bytes to copy (%v) is greater than reader length (%v)", n, r.Len())
+	}
+	for {
+		// Read from the current data reference first.
+		if r.r.Len() > 0 {
+			nCopied, err := io.CopyN(w, r, int64(math.Min(float64(n), float64(r.r.Len()))))
+			n -= nCopied
+			if err != nil {
+				return err
+			}
+		}
+		// Done when there are no bytes left to read.
+		if n == 0 {
+			return nil
+		}
+		// A data reference can be cheaply copied when:
+		// - The writer is at a split point.
+		// - The data reference is a full chunk reference.
+		// - The size of the chunk is less than or equal to the number of bytes left.
+		for w.buf.Len() == 0 && r.dataRefs[0].Hash == "" && r.dataRefs[0].SizeBytes <= n {
+			w.dataRefs[len(w.dataRefs)-1] = r.dataRefs[0]
+			w.dataRefs = append(w.dataRefs, &DataRef{})
+			w.rangeSize += r.dataRefs[0].SizeBytes
+			n -= r.dataRefs[0].SizeBytes
+			r.dataRefs = r.dataRefs[1:]
+		}
```

Comment:
It seems a little strange that this code (outside of both `Reader` and `Writer`) digs into the internals of each and is responsible for maintaining these invariants.  I'm not super concerned since this is encapsulated at a package level, but it's a bit ugly.

Reply:
I definitely agree. A lot of the struggle was trying to get the internal state changes to happen behind an API associated with the writer/reader respectively. The challenge though is that I need to know about internal state in both the writer and reader to know if I can copy a chunk. There is probably a cleaner way to do this, but there are going to be some significant changes to the writer/reader when they are made concurrent, so I just want something I can work with for now and can iterate on later.